### PR TITLE
Refactor navigation helpers

### PR DESCRIPTION
**Pull request by jacwah**
This is my attempt at a piece of #458.

- Code duplication between `previous()` and `next()` have been removed.
- Finding a chapter and rendering it have been split to separate functions.

There are probably tons of potential improvements, but it still feels a lot cleaner than before :blush:.

### CODE REVIEW

**Comment by azerupi** [comment]
The changes look good at first glance. I am going to do a more in-depth review soon.
Thanks for the PR! :)

File: src/renderer/html_handlebars/helpers/navigation.rs [hunk 1]
```diff
@@ -5,165 +5,133 @@ use serde_json;
...
-                        debug!("[*]: Render template");
-                        // Render template
-                        _h.template()
-                          .ok_or_else(|| RenderError::new("Error with the handlebars template"))
-                          .and_then(|t| {
-                            let mut local_rc = rc.with_context(Context::wraps(&previous_chapter)?);
-                            t.render(r, &mut local_rc)
-                        })?;
+                if let Some(previous) = previous {
+                    if let Some(item) = target.find(&base_path, &path, &item, previous)? {
```
Comment (Michael-F-Bryan) [inline src/renderer/html_handlebars/helpers/navigation.rs]: :+1: This looks so much nicer than the massive block we had before!

File: src/renderer/html_handlebars/helpers/navigation.rs [hunk 2]
```diff
@@ -5,165 +5,133 @@ use serde_json;
...
 
+   Ok(None)
+}
+
+fn render(
+    _h: &Helper,
+    r: &Handlebars,
+    rc: &mut RenderContext,
+    chapter: &StringMap,
+) -> Result<(), RenderError> {
```
Comment (Michael-F-Bryan) [inline src/renderer/html_handlebars/helpers/navigation.rs]: How hard would it be to write a test which checks this handlebars helper? Would it be possible to mock up a quick template then run the handlebars helper and assert that the rendered output looks like we expect it to?

File: src/renderer/html_handlebars/helpers/navigation.rs [hunk 3]
```diff
@@ -5,165 +5,133 @@ use serde_json;
...
+    Next,
+}
+
+impl Target {
+    /// Returns target if found.
+    fn find(&self,
+            base_path: &String,
+            current_path: &String,
+            current_item: &StringMap,
+            previous_item: StringMap,
```
[the remark this answers is not in the document]
Comment (jacwah) [reply src/renderer/html_handlebars/helpers/navigation.rs]: I was probably influenced by the fact that the call site already has `previous_item` available. This way we avoid a second clone. When looking at `Target::find` in isolation, that doesn't make a lot of though, as you point out.

**Comment by jacwah** [comment]
@Michael-F-Bryan I changed the signature of `Target::find` and added some tests for the helpers. Do you have any other types of tests in mind?

File: src/renderer/html_handlebars/helpers/navigation.rs [hunk 4]
```diff
@@ -169,8 +168,67 @@ mod tests {
...
+               "path": "three.path"
+            }
+         ]
+      });
+
+      let mut h = Handlebars::new();
+      h.register_helper("previous", Box::new(previous));
+      h.register_helper("next", Box::new(next));
+
+      assert_eq!(
```
Comment (Michael-F-Bryan) [inline src/renderer/html_handlebars/helpers/navigation.rs]: I usually break this out into a couple lines. So at the top of the test I'll assign what I expect to a `should_be` variable, then assign what we actually get to `got`. Then at the bottom you can write `assert_eq!(got, should_be);` (e.g. [config.rs]).

It's definitely a personal taste thing, but it's nice when you can look at the top and figure out what the expected output should be at a glance without needing to look for all `assert_eq!()` calls. 

[config.rs]: https://github.com/rust-lang-nursery/mdBook/blob/fb99276f528307238dc5c1c66f17e6524efcf4ad/src/config.rs#L296-L327

**Comment by Michael-F-Bryan** [comment]
> Do you have any other types of tests in mind?

The tests you've written look good to me. They're clear, concise, and test most common scenarios :+1:

I restarted the Mac travis build. It looks like it was some spurious Travis issue that failed the build and not anything to do with you. Once travis gives the green light I reckon this is ready to merge.

**Comment by Michael-F-Bryan** [comment]
The failed build for OSX/beta is completely unrelated to this PR and seems to be some sort of permissions issue with our `.travis.yml`. I encountered the exact same issue when trying to merge #422 and have opened an issue (#490) to figure out the root cause.

In the meantime I think it's fine to merge this PR. Nice work @jacwah!

**Pull request merged**